charmbits/httpservice/service.go

```diff
@@ -25,47 +23,177 @@ import (
+// associated with the relation in a JSON-marshalable value. This will
+// be unmarshaled and passed as the second setFieldValue argument
+// when the server is started.
+//
+// If the data associated with the relation is incomplete, registerField
+// may return an error with an ErrRelationIncomplete cause,
+// in which case the handler will be stopped until the relation is complete.
+// TODO provide a mechanism for making relations optional.
+//
+// The setFieldValue argument should be a function of the form
+//
+//	func(fv *T, val D) error
+//
+// where T is the struct-field type to be registered and D is the JSON-marshalable
+// type returned from getVal (the data that will be sent to the
+// server). The setFieldValue function will be called within the service whenever the
+// value returned by getVal changes. The value of fv will point to the
+// field to fill in from the relation data. Note that each time the
+// relation changes while the service is still running, the fv value
+// will be the same. If the change in the value requires the service to
+// be restarted, setFieldValue should return an error with an
+// ErrRestartNeeded cause. If any other error is returned, it will
+// be logged as a warning and made available in the charm status.
+// Even if setFieldValue returns an error, it should still set the value
+// appropriately (for example by zeroing out the field).
+func RegisterRelationType(
+	registerField func(r *hook.Registry, tag string) (getVal func() (interface{}, error)),
+	setFieldValue interface{},
+) {
+	registeredRelationsMutex.Lock()
+	defer registeredRelationsMutex.Unlock()
+	reg, err := newRegisteredRelation(registerField, reflect.ValueOf(setFieldValue))
+	if err != nil {
+		panic(errgo.Notef(err, "cannot register relation"))
+	}
+	if _, ok := registeredRelations[reg.t]; ok {
+		panic("type " + reg.t.String() + " registered twice")
+	}
+	registeredRelations[reg.t] = reg
+}
+
+func newRegisteredRelation(
+	registerField func(r *hook.Registry, tag string) (getVal func() (interface{}, error)),
+	setf reflect.Value,
+) (*registeredRelation, error) {
+	sett := setf.Type()
+	if sett.Kind() != reflect.Func {
+		return nil, errgo.Newf("setFieldValue function argument to RegisterRelationType is %v not func", sett)
+	}
+	if sett.NumOut() != 1 {
+		return nil, errgo.Newf("setFieldValue function argument to RegisterRelationType has wrong return count, got %d, want 1", sett.NumOut())
+	}
+	if sett.Out(0) != errorType {
+		return nil, errgo.Newf("setFieldValue function argument to RegisterRelationType has wrong return type, got %s, want error", sett.Out(0))
+	}
+	if sett.NumIn() != 2 {
+		return nil, errgo.Newf("setFieldValue function argument to RegisterRelationType has wrong argument count; got %d, want 2", sett.NumIn())
+	}
+	T := sett.In(0)
+	if T.Kind() != reflect.Ptr {
+		return nil, errgo.Newf("setFieldValue function argument to RegisterRelationType has wrong first argument type; got %s want pointer", T)
```

Comment:
Text could be turned to error constants in case some client might need them. Otherwise leave them.